cdc/processor/pipeline/table_actor.go

```diff
@@ -0,0 +1,351 @@
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/entry"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/redo"
+	"github.com/pingcap/tiflow/cdc/sink"
+	"github.com/pingcap/tiflow/cdc/sink/common"
+	"github.com/pingcap/tiflow/pkg/actor"
+	"github.com/pingcap/tiflow/pkg/actor/message"
+	serverConfig "github.com/pingcap/tiflow/pkg/config"
+	cdcContext "github.com/pingcap/tiflow/pkg/context"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/pipeline"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	_ TablePipeline = (*tableActor)(nil)
+	_ actor.Actor   = (*tableActor)(nil)
+)
+
+type tableActor struct {
+	cancel    context.CancelFunc
+	wg        *errgroup.Group
+	reportErr func(error)
+
+	mb actor.Mailbox
+
+	changefeedID string
+	// quoted schema and table, used in metrics only
+	tableName     string
+	tableID       int64
+	markTableID   int64
+	cyclicEnabled bool
+	memoryQuota   uint64
+	mounter       entry.Mounter
+	replicaInfo   *model.TableReplicaInfo
+	sink          sink.Sink
+	targetTs      model.Ts
+
+	started bool
+	stopped bool
+	err     error
+
+	info             *cdcContext.ChangefeedVars
+	vars             *cdcContext.GlobalVars
+	replConfig       *serverConfig.ReplicaConfig
+	tableActorRouter *actor.Router
+
+	pullerNode *pullerNode
+	sortNode   *sorterNode
+	sinkNode   *sinkNode
+
+	nodes   []*ActorNode
+	actorID actor.ID
+}
+
+// NewTableActor creates a table actor.
+func NewTableActor(cdcCtx cdcContext.Context,
+	mounter entry.Mounter,
+	tableID model.TableID,
+	tableName string,
+	replicaInfo *model.TableReplicaInfo,
+	sink sink.Sink,
+	targetTs model.Ts,
+) (TablePipeline, error) {
+	config := cdcCtx.ChangefeedVars().Info.Config
+	cyclicEnabled := config.Cyclic != nil && config.Cyclic.IsEnabled()
+	info := cdcCtx.ChangefeedVars()
+	vars := cdcCtx.GlobalVars()
+
+	actorID := vars.TableActorSystem.ActorID(info.ID, tableID)
+	mb := actor.NewMailbox(actorID, defaultOutputChannelSize)
+	// Cancel should be able to release all sub-goroutines in the actor.
+	ctx, cancel := context.WithCancel(cdcCtx)
+	// All sub-goroutines should be spawn in the wait group.
+	wg, cctx := errgroup.WithContext(ctx)
+	table := &tableActor{
+		reportErr: cdcCtx.Throw,
+		mb:        mb,
+		wg:        wg,
+		cancel:    cancel,
+
+		tableID:       tableID,
+		markTableID:   replicaInfo.MarkTableID,
+		tableName:     tableName,
+		cyclicEnabled: cyclicEnabled,
+		memoryQuota:   serverConfig.GetGlobalServerConfig().PerTableMemoryQuota,
+		mounter:       mounter,
+		replicaInfo:   replicaInfo,
+		replConfig:    config,
+		sink:          sink,
+		targetTs:      targetTs,
+		started:       false,
+
+		info:             info,
+		vars:             vars,
+		tableActorRouter: vars.TableActorSystem.Router(),
+		actorID:          actorID,
+	}
+
+	log.Info("spawn and start table actor", zap.String("tableName", tableName), zap.Int64("tableID", tableID))
```

Comment:
Better to record the duration between table actor `start` and `start done`.

Reply:
done